missing RESTeasy classes in native build of JAX-RS app

**Describe the bug**
A JAX-RS app which works fine when built as a JAR fails when built as a native executable.

**Expected behavior**
JAX-RS endpoint should work and emit JSON.

**Actual behavior**
JAX-RS endpoint fails during lazy initialization of CDI beans due to missing classes in RESTeasy.

**Additional context**
See reproducer here: https://github.com/emattheis/quarkus-native-issue

hi @emattheis, been looking into this issue a bit and for starters I'll propose a workaround for you.

There seems to be some issue with directly injecting constructs like `Client` instead of using a REST Client interface. I was able to get your sample working by changing the `GreetingService` class to the following:

```java
@ApplicationScoped
public class GreetingService {
    @Inject
    @RestClient
    DummyRestClient dummyRc;

    public Stream<String> getGreetings() {
        dummyRc.ping();
        return Stream.of("Hello, Quarkus!");
    }

    @RegisterRestClient(baseUri = "https://quarkus.io", configKey = "serviceEndpoint")
    public static interface DummyRestClient {
        @HEAD
        public void ping();
    }
}
```

If I go back to the original code, I see that the root issue is the CNFE's -- the "incomplete JSON" error message is a red-herring caused by the try-catch block on the `JsonGenerator`. If we remove the try-catch, we get the usual CNFE's in the `org.jboss.resteasy.plugins.providers.*` package but also get the following NPE:

```
Caused by: java.lang.NullPointerException
	at org.jboss.resteasy.core.ConstructorInjectorImpl.<init>(ConstructorInjectorImpl.java:55)
	at org.jboss.resteasy.core.InjectorFactoryImpl.createConstructor(InjectorFactoryImpl.java:61)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.injectedInstance(ResteasyProviderFactoryImpl.java:1422)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$AbstractInterceptorFactory.createInterceptor(JaxrsInterceptorRegistryImpl.java:151)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.initialize(JaxrsInterceptorRegistryImpl.java:169)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.checkInitialize(JaxrsInterceptorRegistryImpl.java:184)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.getInterceptor(JaxrsInterceptorRegistryImpl.java:195)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$AbstractInterceptorFactory.postMatch(JaxrsInterceptorRegistryImpl.java:139)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl.postMatch(JaxrsInterceptorRegistryImpl.java:284)
	at org.jboss.resteasy.client.jaxrs.internal.ClientConfiguration.getRequestFilters(ClientConfiguration.java:116)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.getRequestFilters(ClientInvocation.java:454)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:671)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:485)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:65)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocationBuilder.head(ClientInvocationBuilder.java:262)
	at io.github.emattheis.quarkus.issues.GreetingService.getGreetings(GreetingService.java:39)
	at io.github.emattheis.quarkus.issues.GreetingService_ClientProxy.getGreetings(GreetingService_ClientProxy.zig:50)
	at io.github.emattheis.quarkus.issues.Greeter.write(Greeter.java:39)
        ... <more stack>
```

--- a/extensions/rest-client/deployment/src/main/java/io/quarkus/restclient/deployment/RestClientProcessor.java
+++ b/extensions/rest-client/deployment/src/main/java/io/quarkus/restclient/deployment/RestClientProcessor.java
@@ -83,6 +83,9 @@ class RestClientProcessor {
     private static final DotName REGISTER_PROVIDER = DotName.createSimple(RegisterProvider.class.getName());
     private static final DotName REGISTER_PROVIDERS = DotName.createSimple(RegisterProviders.class.getName());
 
+    private static final DotName CLIENT_REQUEST_FILTER = DotName.createSimple(ClientRequestFilter.class.getName());
+    private static final DotName CLIENT_RESPONSE_FILTER = DotName.createSimple(ClientResponseFilter.class.getName());
+
     private static final String PROVIDERS_SERVICE_FILE = "META-INF/services/" + Providers.class.getName();
 
     @BuildStep
@@ -340,6 +343,15 @@ class RestClientProcessor {
         for (AnnotationInstance annotationInstance : allInstances) {
             reflectiveClass
                     .produce(new ReflectiveClassBuildItem(false, false, annotationInstance.value().asClass().toString()));
+        }
+
+        // now retain all un-annotated implementations of ClientRequestFilter and ClientResponseFilter
+        // in case they are programmatically registered by applications
+        for (ClassInfo info : index.getAllKnownImplementors(CLIENT_REQUEST_FILTER)) {
+            reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, info.name().toString()));
+        }
+        for (ClassInfo info : index.getAllKnownImplementors(CLIENT_RESPONSE_FILTER)) {
+            reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, info.name().toString()));
         }
     }
 

--- a/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java
+++ b/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java
@@ -20,6 +20,7 @@ import io.quarkus.runtime.annotations.Recorder;
 public class RestClientRecorder {
 
     public static ResteasyProviderFactory providerFactory;
+    public static boolean SSL_ENABLED;
 
     public void setRestClientBuilderResolver() {
         RestClientBuilderResolver.setInstance(new BuilderResolver());

--- a/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java
+++ b/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java
@@ -18,6 +18,9 @@ import io.quarkus.runtime.annotations.Recorder;
 
 @Recorder
 public class RestClientRecorder {
+
+    public static ResteasyProviderFactory providerFactory;
+
     public void setRestClientBuilderResolver() {
         RestClientBuilderResolver.setInstance(new BuilderResolver());
     }
@@ -58,6 +61,7 @@ public class RestClientRecorder {
         }
 
         RestClientBuilderImpl.setProviderFactory(clientProviderFactory);
+        providerFactory = clientProviderFactory;
     }
 
     private static void registerProviders(ResteasyProviderFactory clientProviderFactory, Set<String> providersToRegister,

--- a/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java
+++ b/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java
@@ -23,6 +23,7 @@ public class RestClientRecorder {
     }
 
     public void setSslEnabled(boolean sslEnabled) {
+        SSL_ENABLED = sslEnabled;
         RestClientBuilderImpl.setSslEnabled(sslEnabled);
     }
 

--- a/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/graal/ClientBuilderReplacement.java
+++ b/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/graal/ClientBuilderReplacement.java
@@ -2,6 +2,8 @@ package io.quarkus.restclient.runtime.graal;
 
 import javax.ws.rs.client.ClientBuilder;
 
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
+import org.jboss.resteasy.client.jaxrs.engines.URLConnectionClientEngineBuilder;
 import org.jboss.resteasy.client.jaxrs.internal.LocalResteasyProviderFactory;
 import org.jboss.resteasy.client.jaxrs.internal.ResteasyClientBuilderImpl;
 
@@ -15,7 +17,14 @@ final class ClientBuilderReplacement {
 
     @Substitute
     public static ClientBuilder newBuilder() {
-        return new ResteasyClientBuilderImpl()
-                .providerFactory(new LocalResteasyProviderFactory(RestClientRecorder.providerFactory));
+        ResteasyClientBuilder client = new ResteasyClientBuilderImpl();
+        client.providerFactory(new LocalResteasyProviderFactory(RestClientRecorder.providerFactory));
+        if (!RestClientRecorder.SSL_ENABLED) {
+            client.httpEngine(new URLConnectionClientEngineBuilder().resteasyClientBuilder(client).build());
+            client.sslContext(null);
+            client.trustStore(null);
+            client.keyStore(null, "");
+        }
+        return client;
     }
 }

--- a/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/graal/ClientBuilderReplacement.java
+++ b/extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/graal/ClientBuilderReplacement.java
@@ -2,16 +2,20 @@ package io.quarkus.restclient.runtime.graal;
 
 import javax.ws.rs.client.ClientBuilder;
 
+import org.jboss.resteasy.client.jaxrs.internal.LocalResteasyProviderFactory;
 import org.jboss.resteasy.client.jaxrs.internal.ResteasyClientBuilderImpl;
 
 import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
+
+import io.quarkus.restclient.runtime.RestClientRecorder;
 
 @TargetClass(ClientBuilder.class)
 final class ClientBuilderReplacement {
 
     @Substitute
     public static ClientBuilder newBuilder() {
-        return new ResteasyClientBuilderImpl();
+        return new ResteasyClientBuilderImpl()
+                .providerFactory(new LocalResteasyProviderFactory(RestClientRecorder.providerFactory));
     }
 }

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/ClientProducer.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/ClientProducer.java
@@ -1,0 +1,31 @@
+package io.quarkus.it.rest;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+
+@ApplicationScoped
+public class ClientProducer {
+    private Client client;
+
+    @PostConstruct
+    void init() {
+        client = ClientBuilder.newClient().register(LoggingFilter.class);
+    }
+
+    @PreDestroy
+    void close() {
+        if (client != null) {
+            client.close();
+        }
+    }
+
+    @Produces
+    @ApplicationScoped
+    public Client getClient() {
+        return client;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/ClientResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/ClientResource.java
@@ -6,11 +6,15 @@ import java.util.Map;
 import java.util.concurrent.CompletionStage;
 
 import javax.inject.Inject;
+import javax.inject.Provider;
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 import javax.ws.rs.Produces;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.core.MediaType;
 
 import org.eclipse.microprofile.config.ConfigProvider;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.eclipse.microprofile.rest.client.RestClientBuilder;
 import org.eclipse.microprofile.rest.client.inject.RestClient;
 
@@ -34,6 +38,13 @@ public class ClientResource {
     @Inject
     @RestClient
     RestClientBaseUriConfigKeyInterface restClientBaseUriConfigKeyInterface;
+
+    @Inject
+    @ConfigProperty(name = "loopback/mp-rest/url", defaultValue = "http://localhost:8080/loopback")
+    Provider<String> loopbackEndpoint;
+
+    @Inject
+    Client client;
 
     @GET
     @Path("/annotation/configKey")
@@ -139,5 +150,33 @@ public class ClientResource {
     @Produces("text/plain")
     public String getDefaultInterfaceScope() {
         return Arc.container().instance(RestClientInterface.class, RestClient.LITERAL).getBean().getScope().getName();
+    }
+
+    @GET
+    @Path("/jaxrs-client")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Greeting testJaxrsClient() throws ClassNotFoundException {
+        Greeting greeting = client.target(loopbackEndpoint.get())
+                .request()
+                .get(Greeting.class);
+        // The LoggingFilter should be programmatically registered in io.quarkus.it.rest.ClientProducer.init()
+        if (!client.getConfiguration().isRegistered(Class.forName("io.quarkus.it.rest.LoggingFilter")))
+            throw new IllegalStateException("LoggingFilter should be registered on injected Client");
+        if (getFilterCount() != 2)
+            throw new IllegalStateException("Call count should have been 2 but was " + getFilterCount());
+        return greeting;
+    }
+
+    private int getFilterCount() {
+        try {
+            // Must use reflection to check filter call count to ensure that
+            // completely decoupled filters are not removed in native mode
+            return Class.forName("io.quarkus.it.rest.LoggingFilter")
+                    .getDeclaredField("CALL_COUNT")
+                    .getInt(null);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return -1;
+        }
     }
 }

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/Greeting.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/Greeting.java
@@ -1,0 +1,26 @@
+package io.quarkus.it.rest;
+
+import java.time.LocalDate;
+
+import javax.json.bind.annotation.JsonbCreator;
+import javax.json.bind.annotation.JsonbProperty;
+
+public class Greeting {
+
+    private final String message;
+    private final LocalDate date;
+
+    @JsonbCreator
+    public Greeting(@JsonbProperty("message") String message, @JsonbProperty("date") LocalDate date) {
+        this.message = message;
+        this.date = date;
+    }
+
+    public String getMessage() {
+        return message;
+    }
+
+    public LocalDate getDate() {
+        return date;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/LoggingFilter.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/LoggingFilter.java
@@ -1,0 +1,33 @@
+package io.quarkus.it.rest;
+
+import java.io.IOException;
+
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+import javax.ws.rs.client.ClientResponseContext;
+import javax.ws.rs.client.ClientResponseFilter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoggingFilter
+        implements ClientRequestFilter, ClientResponseFilter {
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public static int CALL_COUNT = 0;
+
+    @Override
+    public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext)
+            throws IOException {
+        CALL_COUNT++;
+        log.info("<< {} {}", responseContext.getStatusInfo().getStatusCode(),
+                responseContext.getStatusInfo().getReasonPhrase());
+    }
+
+    @Override
+    public void filter(ClientRequestContext requestContext)
+            throws IOException {
+        CALL_COUNT++;
+        log.info(">> {} {}", requestContext.getMethod(), requestContext.getUri());
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/LoopbackResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/LoopbackResource.java
@@ -1,0 +1,19 @@
+package io.quarkus.it.rest;
+
+import java.time.LocalDate;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/loopback")
+public class LoopbackResource {
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Greeting hello() {
+        return new Greeting("hello self", LocalDate.of(2020, 02, 13));
+    }
+
+}

--- a/integration-tests/main/src/main/resources/application.properties
+++ b/integration-tests/main/src/main/resources/application.properties
@@ -3,6 +3,7 @@ io.quarkus.it.rest.RestInterface/mp-rest/scope=javax.inject.Singleton
 io.quarkus.it.rest.RestClientInterface/mp-rest/url=${test.url}
 restClientConfigKey/mp-rest/url=${test.url}
 restClientBaseUriConfigKey/mp-rest/url=${test.url}
+loopback/mp-rest/url=${test.url}/loopback
 
 org.eclipse.microprofile.rest.client.propagateHeaders=header-name
 # Disabled by default as it establishes external connections.

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/RestClientTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/RestClientTestCase.java
@@ -1,5 +1,7 @@
 package io.quarkus.it.main;
 
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.Matchers.*;
 
 import java.util.List;
@@ -116,6 +118,16 @@ public class RestClientTestCase {
 
         Assertions.assertEquals("javax.inject.Singleton", responseWithSingletonScope);
         Assertions.assertEquals("javax.enterprise.context.Dependent", responseWithDefaultScope);
+    }
+
+    @Test
+    void testJaxrsClientWithFilters() {
+        given()
+                .when().get("/client/jaxrs-client")
+                .then()
+                .statusCode(200)
+                .body(containsString("hello"))
+                .body(containsString("2020-02-13"));
     }
 
     @Test

